apps/arweave/src/ar_kv.erl

```diff
@@ -2,659 +2,632 @@
 
 -behaviour(gen_server).
 
--export([start_link/0, open/2, open/3, open/4, put/3, get/2, get_next_by_prefix/4, get_next/2,
-		get_prev/2, get_range/2, get_range/3, delete/2, delete_range/3, count/1]).
+-export([
+	start_link/0, create_ets/0, open/2, open/3, open/4, put/3, get/2,
+	get_next_by_prefix/4, get_next/2, get_prev/2, get_range/2, get_range/3,
+	delete/2, delete_range/3, count/1
+]).
 
 -export([init/1, handle_cast/2, handle_call/3, handle_info/2, terminate/2]).
 
 -include_lib("arweave/include/ar.hrl").
 -include_lib("arweave/include/ar_config.hrl").
 -include_lib("eunit/include/eunit.hrl").
 
--record(state, {}).
+-define(WITH_DB(Name, Callback), with_db(Name, ?FUNCTION_NAME, Callback)).
+-define(WITH_ITERATOR(Name, IteratorOptions, Callback), with_iterator(Name, ?FUNCTION_NAME, IteratorOptions, Callback)).
+
+-define(DEFAULT_ROCKSDB_DATABASE_OPTIONS, #{
+	create_if_missing => true,
+	create_missing_column_families => true,
+
+	%% these are default values, but they must not be overriden;
+	%% otherwise the syncWAL will not work.
+	allow_mmap_reads => false,
+	allow_mmap_writes => false
+}).
+
+-record(db, {
+	%% name may be undefined in short intervals before opening the database,
+	%% or reopening the database (which implies close and open operations).
+	%% It may happen in case of opening the database with column families.
+	%% NB: records with undefined db_handle must not be stored in the ETS table.
+	name :: term() | undefined,
+	filepath :: file:filename_all(),
+	db_options :: rocksdb:db_options(),
+	%% db_handle may be undefined in short intervals before opening the database,
+	%% or reopening the database (which implies close and open operations).
+	%% NB: records with undefined db_handle must not be stored in the ETS table.
+	db_handle :: rocksdb:db_handle() | undefined,
+
+	%% column families only fields, must be set to undefined for plain databases.
+	cf_names = undefined :: [term()],
+	cf_descriptors = undefined :: [rocksdb:cf_descriptor()],
+	cf_handle = undefined :: rocksdb:cf_handle()
+}).
+
+-define(msg_trigger_timer(Kind, Secret), {msg_trigger_timer, Kind, Secret}).
+-define(msg_trigger_db_flush(Secret), ?msg_trigger_timer(db_flush, Secret)).
+-define(msg_trigger_wal_sync(Secret), ?msg_trigger_timer(wal_sync, Secret)).
+
+-record(timer, {
+	interval_ms :: pos_integer(),
+	ref :: erlang:reference() | undefined,
+	secret :: erlang:reference() | undefined
```

Comment:
The `secret` is to prevent an external actor from forcing a flush connecting to the arweave instance and sending a message/

Reply:
got it,. It's an extra check just to prevent some possible but very unlikely case - but worth it since this whole PR is about reliability so stamping out any source of corruption is important. Makes sense.